docs(cn): translate learn/you-might-not-need-an-effect into Chinese (#1155)
reactjs/zh-hans.react.dev: src/content/learn/you-might-not-need-an-effect.md

---
title: '你可能不需要 Effect'
---

<Intro>

> Effect 是 React 范式中的一个逃脱方案。它们让你可以 “逃出” React 并使组件和一些外部系统同步，比如非 React 组件、网络和浏览器 DOM。如果没有涉及到外部系统（例如，你想根据 props 或 state 的变化来更新一个组件的 state），你就不应该使用 Effect。移除不必要的 Effect 可以让你的代码更容易理解，运行得更快，并且更少出错。

Reviewer: ```suggestion
Effect 是 React 范式中的一个逃脱方案。它们让你可以“逃出” React 并使组件和一些外部系统同步，比如非 React 组件、网络和浏览器 DOM。如果没有涉及到外部系统（例如，你想根据 props 或 state 的变化来更新一个组件的 state），你就不应该使用 Effect。移除不必要的 Effect 可以让你的代码更容易理解，运行得更快，并且更少出错。
```

</Intro>

<YouWillLearn>

* 为什么以及如何从你的组件中移除 Effect
* 如何在没有 Effect 的情况下缓存昂贵的计算
* 如何在没有 Effect 的情况下重置和调整组件的 state
* 如何在事件处理函数之间共享逻辑
* 应该将哪些逻辑移动到事件处理函数中
* 如何将发生的变动通知到父组件

</YouWillLearn>

## 如何移除不必要的 Effect {/*how-to-remove-unnecessary-effects*/}

有两种不必使用 Effect 的常见情况：

> * **你不必使用 Effect 来转换渲染所需的数据**。例如，你想在展示一个列表前先做筛选。你的直觉可能是写一个当列表变化时更新 state 变量的 Effect。然而，这是低效的。当你更新这个 state 时，React 首先会调用你的组件函数来计算应该显示在屏幕上的内容。然后 React 会把这些变化 “[提交](/learn/render-and-commit)” 到 DOM 中来更新屏幕。然后 React 会执行你的 Effect。如果你的 Effect **也** 立即更新了这个 state，就会重新执行整个流程。为了避免不必要的渲染流程，应在你的组件顶层转换数据。这些代码会在你的 props 或 state 变化时自动重新执行。

Reviewer: ```suggestion
* **你不必使用 Effect 来转换渲染所需的数据**。例如，你想在展示一个列表前先做筛选。你的直觉可能是写一个当列表变化时更新 state 变量的 Effect。然而，这是低效的。当你更新这个 state 时，React 首先会调用你的组件函数来计算应该显示在屏幕上的内容。然后 React 会把这些变化“[提交](/learn/render-and-commit)”到 DOM 中来更新屏幕。然后 React 会执行你的 Effect。如果你的 Effect 也立即更新了这个 state，就会重新执行整个流程。为了避免不必要的渲染流程，应在你的组件顶层转换数据。这些代码会在你的 props 或 state 变化时自动重新执行。
```

> * **你不必使用 Effect 来处理用户事件**。例如，你想在用户购买一个产品时发送一个 `/api/buy` 的 POST 请求并展示一个提示。在这个购买按钮的 click 事件处理函数中，你确切的知道会发生什么。但是你不知道用户做了什么导致 Effect 被执行（例如，点击了某个按钮）。这就是为什么你通常应该在相应的事件处理函数中处理用户事件。

Reviewer: ```suggestion
* **你不必使用 Effect 来处理用户事件**。例如，你想在用户购买一个产品时发送一个 `/api/buy` 的 POST 请求并展示一个提示。在这个购买按钮的点击事件处理函数中，你确切地知道会发生什么。但是你不知道用户做了什么导致 Effect 被执行（例如，点击了某个按钮）。这就是为什么你通常应该在相应的事件处理函数中处理用户事件。
```

> * **你不必使用 Effect 来处理用户事件**。例如，你想在用户购买一个产品时发送一个 `/api/buy` 的 POST 请求并展示一个提示。在这个购买按钮的 click 事件处理函数中，你确切的知道会发生什么。而当一个 Effect 执行时，你不知道用户做过什么（例如，哪个按钮被点击了）。这就是为什么你通常应该在相应的事件处理函数中处理用户事件。

Reviewer 2: > 而当一个 Effect 执行时，你不知道用户做过什么（例如，哪个按钮被点击了）。

翻译成类似下面这样会不会更容易理解一点？

但是你不知道用户做了什么导致 Effect 被执行（例如点击了某个按钮）。

你 **的确** 可以使用 Effect 来和外部系统 [同步](/learn/synchronizing-with-effects#what-are-effects-and-how-are-they-different-from-events) 。例如，你可以写一个 Effect 来保持一个 jQuery 的组件和 React state 之间的同步。你也可以使用 Effect 来获取数据：例如，你可以同步当前的查询搜索和查询结果。请记住，比起直接在你的组件中写 Effect，现代 [框架](/learn/start-a-new-react-project#production-grade-react-frameworks) 提供了更加高效的，内置的数据获取机制。

为了帮助你获得正确的直觉，让我们来看一些常见的实例吧！

### 根据 props 或 state 来更新 state {/*updating-state-based-on-props-or-state*/}

假设你有一个包含了两个 state 变量的组件：`firstName` 和 `lastName`。你想通过把它们联结起来计算出 `fullName`。此外，每当 `firstName` 和 `lastName` 变化时，你希望 `fullName` 都能更新。你的第一直觉可能是添加一个 state 变量：`fullName`，并在一个 Effect 中更新它：

```js {5-9}
function Form() {

>   const [firstName, setFirstName] = useState('泰勒');

Reviewer: 代码块中非注释的英文就不翻译吧？

Reviewer: 纠正一下，代码块中出现的内容（如 `alert`、`console.log` 中的内容）也需要翻译。
此处的字符串 `'Taylor'` 更多是一种作为初始变量传入 `useState`，保留英文比较好

}
```

> 就像之前的例子一样，这即没有必要，也很低效。首先，移除 state 和 Effect：

Reviewer: ```suggestion
就像之前的例子一样，这既没有必要，也很低效。首先，移除 state 和 Effect：
```

```js {3-4}
function TodoList({ todos, filter }) {
  const [newTodo, setNewTodo] = useState('');
  // ✅ 如果 getFilteredTodos() 的耗时不长，这样写就可以了。
  const visibleTodos = getFilteredTodos(todos, filter);
  // ...
}
```

> 一般来说，这段代码没有问题！但是，`getFilteredTodos()` 的耗时可能会很长，或者你有很多 `todos`。这些情况下，当 `newTodo` 这样不相关的 state 变量变化时，你并不想重新执行 `getFilteredTodos()`。

Reviewer: ```suggestion
一般来说，这段代码没有问题！但是，`getFilteredTodos()` 的耗时可能会很长，或者 `todos` 的内容很多。这些情况下，当 `newTodo` 这样不相关的 state 变量变化时，你并不想重新执行 `getFilteredTodos()`。
```

Author: 这里我是故意翻译的“或者你有很多 `todos`。”，原文是“or you have a lot of `todos`.”
这里有点幽默的哈哈，你有很多todos，就是说或者你有很多要做的“事儿”。有点双关的寓意，很好玩。

> 你可以通过 [`useMemo`](/reference/react/useMemo) 这个 Hook 来缓存（或者说 ["memoize"](https://en.wikipedia.org/wiki/Memoization)）一个昂贵的计算。

Reviewer: ```suggestion
你可以使用 [`useMemo`](/reference/react/useMemo) Hook 缓存（或者说 [记忆（memoize）](https://en.wikipedia.org/wiki/Memoization)）一个昂贵的计算。
```

```js {5-8}
import { useMemo, useState } from 'react';

function TodoList({ todos, filter }) {
  const [newTodo, setNewTodo] = useState('');
  const visibleTodos = useMemo(() => {
    // ✅ 除非 todos 或 filter 发生变化，否则不会重新执行
    return getFilteredTodos(todos, filter);
  }, [todos, filter]);
  // ...
}
```

> 或者，写成一行：

Reviewer 2: ```suggestion
或者写成一行：
```

```js {5-6}
import { useMemo, useState } from 'react';

function TodoList({ todos, filter }) {
  const [newTodo, setNewTodo] = useState('');
  // ✅ 除非 todos 或 filter 发生变化，否则不会重新执行 getFilteredTodos()
  const visibleTodos = useMemo(() => getFilteredTodos(todos, filter), [todos, filter]);
  // ...
}
```

**这会告诉 React，除非 `todos` 或 `filter` 发生变化，否则不要重新执行传入的函数**。React 会在初次渲染的时候记住 `getFilteredTodos()` 的返回值。在下一次渲染中，它会检查 `todos` 或 `filter` 是否发生了变化。如果它们跟上次渲染时一样，`useMemo` 会直接返回它最后保存的结果。如果不一样，React 将再次调用传入的函数（并保存它的结果）。

你传入 [`useMemo`](/reference/react/useMemo) 的函数会在渲染期间执行，所以仅在它是 [纯函数](/learn/keeping-components-pure) 时有效。

<DeepDive>

#### 如何判断计算是昂贵的？ {/*how-to-tell-if-a-calculation-is-expensive*/}

> 除非你创建或循环遍历了成千上万个对象，一般来说，它可能不会太耗费时间。如果你想确认一下，可以添加控制台输出来测量某一段代码的执行时间：

Reviewer 2: ```suggestion
一般来说只有你创建或循环遍历了成千上万个对象时才会很耗费时间。如果你想确认一下，可以添加控制台输出来测量某一段代码的执行时间：
```

```js {1,3}
console.time('筛选数组');
const visibleTodos = getFilteredTodos(todos, filter);
console.timeEnd('筛选数组');
```

触发要测量的交互（例如，在输入框中输入）。你会在控制台中看到类似 `筛选数组：0.15ms` 这样的输出日志。如果总耗时达到了一定量级（比方说 `1ms` 或更多），那么把计算结果记忆（memoize）起来可能是有意义的。做一个实验，你可以把计算传入 `useMemo` 中来验证该交互导致的总耗时是减少了还是没什么变化：

```js
console.time('筛选数组');
const visibleTodos = useMemo(() => {
  return getFilteredTodos(todos, filter); // 如果 todos 或 filter 没有发生变化将跳过执行
}, [todos, filter]);
console.timeEnd('筛选数组');
```

`useMemo` 不会让 **第一次** 渲染变快。它只是帮助你跳过不必要的更新。

请注意，你的设备性能可能比用户的更好，因此最好通过人工限制工具来测试性能。例如，Chrome 提供了 [CPU 节流](https://developer.chrome.com/blog/new-in-devtools-61/#throttling) 工具。

> 同时需要注意，在开发环境测试性能不能得到最准确的结果（例如，当开启 [严格模式](/reference/react/StrictMode) 时，你会看到每个组件渲染了两次，而不是一次）。所以为了得到最准确的时间，需要将你的应用构建为生产模式，同时使用与你的用户性能相当的设备进行测试。

Reviewer 2: ```suggestion
同时需要注意的是，在开发环境测试性能并不能得到最准确的结果（例如，当开启 [严格模式](/reference/react/StrictMode) 时，你会看到每个组件渲染了两次，而不是一次）。所以为了得到最准确的时间，需要将你的应用构建为生产模式，同时使用与你的用户性能相当的设备进行测试。
```

</DeepDive>

> ### 当 prop 变化时重置所有 state {/*resetting-all-state-when-a-prop-changes*/}

Reviewer: 除非原文明确表示 prop 仅为“一个”，并且最好是明确具体是哪一个 prop，否则使用 props。
如此处，原文想要表达的意思 prop(s) 改变了就重置 state，虽然原文使用的是单数，但是此处“一个 prop”，或者“这个 prop”的意思非常微弱，因此此处建议使用 props。
其他地方类似。

`ProfilePage` 组件接收一个 prop：`userId`。页面上有一个评论输入框，你用了一个 state：`comment` 来保存它的值。有一天，你发现了一个问题：当你从一个人的个人资料导航到另一个时，`comment` 没有被重置。这导致很容易不小心把评论发送到不正确的个人资料。为了解决这个问题，你想在 `userId` 变化时，清除 `comment` 变量：

```js {4-7}
export default function ProfilePage({ userId }) {
  const [comment, setComment] = useState('');

  // 🔴 避免：当 prop 变化时，在 Effect 中重置 state
  useEffect(() => {
    setComment('');
  }, [userId]);
  // ...
}
```

> 但这是低效的，因为 `ProfilePage` 和它的子组件首先会用旧的值渲染，然后再（用新的值）重新渲染。并且这样做也很复杂，因为你需要在 `ProfilePage` 里面 **所有** 具有 state 的组件中都写这样的代码。例如，如果评论区的 UI 是嵌套的，你可能也想清除嵌套的 comment state。

Reviewer: ```suggestion
但这是低效的，因为 `ProfilePage` 和它的子组件首先会用旧值渲染，然后再用新值重新渲染。并且这样做也很复杂，因为你需要在 `ProfilePage` 里面 **所有** 具有 state 的组件中都写这样的代码。例如，如果评论区的 UI 是嵌套的，你可能也想清除嵌套的 comment state。
```

}
```

像这样 [存储前序渲染的信息](/reference/react/useState#storing-information-from-previous-renders) 可能很难理解，但它比在 Effect 中更新这个 state 要好。上面的例子中，在渲染过程中直接调用了 `setSelection`。当它执行到 `return` 语句退出后，React 将 **立即** 重新渲染 `List`。此时 React 还没有渲染 `List` 的子组件或更新 DOM，这使得 `List` 的子组件可以跳过渲染旧的 `selection` 值。

在渲染期间更新组件时，React 会丢弃已经返回的 JSX 并立即尝试重新渲染。为了避免非常缓慢的级联重试，React 只允许在渲染期间更新 **同一** 组件的状态。如果你在渲染期间更新另一个组件的状态，你会看到一条报错信息。条件判断 `items !== prevItems` 是必要的，它可以避免无限循环。你可以像这样调整 state，但任何其他副作用（比如变化 DOM 或设置的延时）应该留在事件处理函数或 Effect 中，以 [保持组件纯粹](/learn/keeping-components-pure)。

> **虽然这种方式比 Effect 更高效，但大多数组件也不需要它**。无论你怎么做，根据 props 或其他 state 来调整 state 都会使数据流更难理解和调试。总是检查是否可以通过添加 [key 来重置所有 state](#resetting-all-state-when-a-prop-changes)，或者 [在渲染期间计算所需内容](#updating-state-based-on-props-or-state)。例如，你可以存储已选中的 **项目ID** 而不是存储（并重置）已选中的 **项目：**

Reviewer: ```suggestion
**虽然这种方式比 Effect 更高效，但大多数组件也不需要它**。无论你怎么做，根据 props 或其他 state 来调整 state 都会使数据流更难理解和调试。总是检查是否可以通过添加 [key 来重置所有 state](#resetting-all-state-when-a-prop-changes)，或者 [在渲染期间计算所需内容](#updating-state-based-on-props-or-state)。例如，你可以存储已选中的 **item ID** 而不是存储（并重置）已选中的 **item**：
```

}
```

这既移除了不必要的 Effect，又修复了问题。

### 发送 POST 请求 {/*sending-a-post-request*/}

> 这个 `Form` 组件会发送两种 POST 请求。它在页面加载之际会发送一个分析请求。当你填写表格并点击 `提交` 按钮时，它会向 `/api/register` 接口发送一个 POST 请求：

Reviewer: ```suggestion
这个 `Form` 组件会发送两种 POST 请求。它在页面加载之际会发送一个分析请求。当你填写表格并点击提交按钮时，它会向 `/api/register` 接口发送一个 POST 请求：
```
